kernel/aster-nix/src/syscall/mount.rs

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use super::SyscallReturn;
+use crate::{
+    fs::{
+        exfat::{ExfatFS, ExfatMountOptions},
+        ext2::Ext2,
+        fs_resolver::{FsPath, AT_FDCWD},
+        path::Dentry,
+        utils::{FileSystem, InodeType},
+    },
+    prelude::*,
+    syscall::constants::MAX_FILENAME_LEN,
+    util::read_cstring_from_user,
+};
+
+/// The data argument is interpreted by the different filesystems.
+/// Typically it is a string of comma-separated options understood by
+/// this filesystem. The current implementation only considers the case
+/// where it is NULL. Because it should be interpreted by the specific filesystems.
+pub fn sys_mount(
+    devname_addr: Vaddr,
+    dirname_addr: Vaddr,
+    fstype_addr: Vaddr,
+    flags: u64,
+    data: Vaddr,
+) -> Result<SyscallReturn> {
+    let devname = read_cstring_from_user(devname_addr, MAX_FILENAME_LEN)?;
+    let dirname = read_cstring_from_user(dirname_addr, MAX_FILENAME_LEN)?;
+    let mount_flags = MountFlags::from_bits_truncate(flags as u32);
+    debug!(
+        "devname = {:?}, dirname = {:?}, fstype = 0x{:x}, flags = {:?}, data = 0x{:x}",
+        devname, dirname, fstype_addr, mount_flags, data,
+    );
+
+    let current = current!();
+    let target_dentry = {
+        let dirname = dirname.to_string_lossy();
+        if dirname.is_empty() {
+            return_errno_with_message!(Errno::ENOENT, "dirname is empty");
+        }
+        let fs_path = FsPath::new(AT_FDCWD, dirname.as_ref())?;
+        current.fs().read().lookup(&fs_path)?
+    };
+
+    if mount_flags.contains(MountFlags::MS_REMOUNT) && mount_flags.contains(MountFlags::MS_BIND) {
+        do_reconfigure_mnt()?;
+    } else if mount_flags.contains(MountFlags::MS_REMOUNT) {
+        do_remount()?;
+    } else if mount_flags.contains(MountFlags::MS_BIND) {
+        do_loopback(
+            devname.clone(),
+            target_dentry.clone(),
+            mount_flags.contains(MountFlags::MS_REC),
+        )?;
+    } else if mount_flags.contains(MountFlags::MS_SHARED)
+        | mount_flags.contains(MountFlags::MS_PRIVATE)
+        | mount_flags.contains(MountFlags::MS_SLAVE)
+        | mount_flags.contains(MountFlags::MS_UNBINDABLE)
+    {
+        do_change_type()?;
+    } else if mount_flags.contains(MountFlags::MS_MOVE) {
+        do_move_mount_old(devname, target_dentry)?;
+    } else {
+        do_new_mount(devname, fstype_addr, target_dentry)?;
+    }
+
+    Ok(SyscallReturn::Return(0))
+}
+
+fn do_reconfigure_mnt() -> Result<()> {
+    todo!()
+}
+
+fn do_remount() -> Result<()> {
+    todo!()
+}
+
+/// Bind a mount to a new location.
+fn do_loopback(old_name: CString, new_dentry: Arc<Dentry>, recursive: bool) -> Result<()> {
```

Comment:
Thank you for the feedback. I have updated the function argument to use "source" and "destination" instead of "old" and "new" to better reflect their purpose in the context of mount operations.